File: server/skills/events.py
from dataclasses import dataclass
from enum import StrEnum


class EventType(StrEnum):
    THINKING = "thinking"   # Copilot reasoning / status narration
    SKILL    = "skill"      # Skill being invoked
    CONTENT  = "content"    # Streaming LLM token output
    WAITING  = "waiting"    # Async operation in progress (DQL execution etc.)
    RESULT   = "result"     # Final structured output from a skill
    TABLE    = "table"      # JSON-encoded query results for collapsible table UI
    SUGGEST       = "suggest"        # Downstream skill suggestion offered to user
    ERROR         = "error"          # Something went wrong
    DONE          = "done"           # Stream complete
    TRANSCRIPTION = "transcription"  # Voice input transcribed to text
    CLARIFY       = "clarify"        # Copilot needs more info — JSON payload with options
    BLOCKED       = "blocked"        # Request blocked by guardrail


@dataclass
class CopilotEvent:
    type: EventType
    content: str

    def to_sse(self) -> dict:
        """Serialize to SSE-compatible dict for sse-starlette."""
        return {
            "event": self.type.value,
            "data": self.content,
        }

File: widget/src/types/events.ts
export type EventType =
  | 'thinking'
  | 'skill'
  | 'content'
  | 'waiting'
  | 'result'
  | 'table'
  | 'suggest'
  | 'error'
  | 'done'
  | 'transcription'
  | 'clarify'
  | 'blocked'

export interface CopilotEvent {
  type: EventType
  content: string
}

export interface Suggestion {
  skillId: string
  label: string
}

export interface PhaseTimes {
  skill?:   number   // ms timestamp when skill event arrived
  waiting?: number   // ms timestamp when waiting event arrived (query executing)
  result?:  number   // ms timestamp when result event arrived
  done?:    number   // ms timestamp when done event arrived
}

export interface Message {
  id: string
  role: 'user' | 'copilot'
  text?: string
  originalQuery?: string   // original user query, used for copy-with-context
  events: CopilotEvent[]
  suggestions: Suggestion[]
  isStreaming: boolean
  startedAt: number        // Date.now() when query was sent
  phaseTimes: PhaseTimes
}
